realm-jni/src/io_realm_internal_table.cpp

```diff
@@ -1484,3 +1484,33 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_Table_nativeSetPrimaryKey(
     } CATCH_STD()
     return 0;
 }
+
+// Fixes interop issue with Cocoa Realm where the Primary Key table had different types.
+// This effects:
+// - All Realms created by Cocoa and used by Realm-android up to 0.80.1
+// - All Realms created by Realm-Android 0.80.1 and below
+// See https://github.com/realm/realm-java/issues/1059
+JNIEXPORT void JNICALL Java_io_realm_internal_Table_nativeMigratePrimaryKeyTableIfNeeded
+    (JNIEnv* env, jobject, jlong groupNativePtr, jlong privateKeyTableNativePtr)
+{
+    Group* group = G(groupNativePtr);
+    Table* pk_table = TBL(privateKeyTableNativePtr);
+    if (pk_table->get_column_type(io_realm_internal_Table_PRIMARY_KEY_FIELD_COLUMN_INDEX) == type_Int) {
+        StringData tmp_col_name = StringData("tmp_field_name");
+        std::size_t tmp_col_ndx = pk_table->add_column(DataType(type_String), tmp_col_name);
+
+        // Create tmp string column with field name instead of column index
+        int rows = pk_table->size();
+        for (int row_ndx = 0; row_ndx < rows; row_ndx++) {
+            StringData table_name = pk_table->get_string(io_realm_internal_Table_PRIMARY_KEY_CLASS_COLUMN_INDEX, row_ndx);
+            int64_t col_ndx = pk_table->get_int(io_realm_internal_Table_PRIMARY_KEY_FIELD_COLUMN_INDEX, row_ndx);
```

Comment:
You have to be careful here: an integer column is always `int64_t` but column index is `size_t` (which is an `unsigned long`). They differ on a 32 bit architecture.